nuxeo-core/nuxeo-core/src/main/java/org/nuxeo/ecm/core/model/DocumentModelResolver.java

```diff
@@ -175,24 +174,29 @@ public Object fetch(Object value) throws IllegalStateException {
             REF ref = REF.fromValue((String) value);
             if (ref != null) {
                 try (CloseableCoreSession session = CoreInstance.openCoreSession(ref.repo)) {
-                    try {
-                        DocumentModel doc;
-                        switch (mode) {
-                        case ID_REF:
-                            doc = session.getDocument(new IdRef(ref.ref));
-                            break;
-                        case PATH_REF:
-                            doc = session.getDocument(new PathRef(ref.ref));
-                            break;
-                        default:
-                            throw new UnsupportedOperationException();
-                        }
-                        // detach because we're about to close the session
-                        doc.detach(true);
-                        return doc;
-                    } catch (DocumentNotFoundException e) {
+                    DocumentModel doc;
+                    DocumentRef docRef;
+
+                    switch (mode) {
+                    case ID_REF:
+                        docRef = new IdRef(ref.ref);
+                        break;
+                    case PATH_REF:
+                        docRef = new PathRef(ref.ref);
+                        break;
+                    default:
+                        throw new UnsupportedOperationException();
+                    }
+
+                    if (!session.exists(docRef)) {
+                        // the document doesn't exist or is not accessible by the current user
                         return null;
                     }
+
+                    doc = session.getDocument(docRef);
+                    // detach because we're about to close the session
+                    doc.detach(true);
+                    return doc;
```

Comment:
it is cleaner this way, but we have an addition call to session.exists that is done each time even for accessible documents could impact perf if this method is a hotspot

Reply:
@bdelbosc we could be back to the original code and just catch `DocumentSecurityException` and `DocumentNotFoundException`. I don't think the extra internal `resolveReference` that returns a `Document` is a big issue though, as it access cached data after the first call. @bdelbosc your call

Reply:
yes I am ok, please merge the initial PR

Reply:
Merging then